### PR TITLE
Step Functions Infrastructure, Roles, and Policies

### PR DESCRIPTION
**Pull request by queueburt**
Changes to **metaflow-cfn-template.yml:**

Role and policy additions:
- Enable EventBridge and Step Functions access for Metaflow Isolated User
- Create Role for Step Functions to access Event Bridge, S3, DynamoDB, and Batch
- Create Role for Event Bridge to trigger Step Functions
- Modify Batch role to allow access to DynamoDB

Infrastructure Additions:
- DynamoDB for Step Functions state management

Fixes:
- Per feedback, creating an ECSInstanceProfile role that's specific to each stack's Batch deployment as opposed to relying on Batch's [ecsInstanceRole](https://docs.aws.amazon.com/batch/latest/userguide/instance_IAM_role.html).
- Also per feedback, making Sagemaker Notebooks an optional resource.

Addition of **metaflow-cfn-ONLY-stepfunctions.yml:**

- Addition of a template that only includes DynamoDB and respective roles.  **USE WITH CARE.  This will require modification of existing roles by hand to enable access to DynamoDB from Batch and Batch from Step Functions.**

### CODE REVIEW

**Comment by queueburt** [comment]
I'm feeling like I no longer like the idea of a separate CFN template in this PR to incorporate Step Functions in a standalone way.  The only users it helps are manual deployers who suddenly want to use CFN for Step Functions, and it means we would now have two places to maintain parity for new features.  @savingoyal I think I'm going to remove **metaflow-cfn-ONLY-stepfunctions.yml** from this PR (because if I don't I have to make a bunch of adjustments to support DB migration), but I wanted to float it out to the larger team before I do.

**Comment by savingoyal** [comment]
@queueburt Yes, let's remove **metaflow-cfn-ONLY-stepfunctions.yml**

File: aws/cloudformation/metaflow-cfn-template.yml [hunk 1]
```diff
@@ -263,6 +271,26 @@ Resources:
               Condition:
                 StringNotEquals:
                   s3:authType: REST-HEADER
+  ECSInstanceProfile:
+    Type: AWS::IAM::InstanceProfile
+    Properties:
+      Path: /
+      Roles:
+        - !Ref ECSInstanceRole
+  ECSInstanceRole:
```
Comment (tobias-gp) [inline aws/cloudformation/metaflow-cfn-template.yml:280]: Great, that's also something we missed :)

File: aws/cloudformation/metaflow-cfn-template.yml [hunk 2]
```diff
@@ -29,16 +29,16 @@ Parameters:
...
     Description: 'Maximum VCPUs for Batch Compute Environment [2-16]'
   DesiredVCPUBatch: 
     Type: Number
     Default: 8
-    MinValue: 2
+    MinValue: 0
     MaxValue: 16
-    AllowedValues: [2,4,8,16]
+    AllowedValues: [0,2,4,8,16]
     Description: 'Desired Starting VCPUs for Batch Compute Environment [2-16]'
```
Comment (tobias-gp) [inline aws/cloudformation/metaflow-cfn-template.yml:42]: The textual description must be updated :)

**Comment by tobias-gp** [comment]
I just deployed the stack via CF, and it deploys just fine :) I haven't tested the SF functionality though.

File: aws/cloudformation/metaflow-cfn-template.yml (hunk 2 above)
Comment (queueburt) [reply aws/cloudformation/metaflow-cfn-template.yml]: Good catch.  Fixed in a recent commit.

File: aws/cloudformation/metaflow-cfn-template.yml [hunk 3]
```diff
@@ -315,6 +372,88 @@ Resources:
...
+            
+            upgrade_patch = request.Request(upgrade_endpoint, method='PATCH')
+            with request.urlopen(upgrade_patch) as upgrade:
+              response['upgrade-result'] = upgrade.read().decode()
+              
+            with request.urlopen(status_endpoint) as status:
+              response['final-status'] = json.loads(status.read())
+          
+            print(response)
+            return(response)
```
Comment (queueburt) [inline aws/cloudformation/metaflow-cfn-template.yml:447]: @savingoyal @ferras This Lambda Function is probably deserving of a quick review to make sure its functionality is acceptable for a user-facing workflow.  Basically it's just a return of "initial status", "upgrade result", and "final status" printed to Cloudwatch for posterity.

**Comment by queueburt** [comment]
Final update: health-check port for the new traffic flow has been changed to 8080, and is passing in my test deployments.  I'm comfortable calling this good-for-merge assuming some +1 love from @savingoyal @tobias-gp @ferras and @tuulos

**Pull request merged**